app/src/org/commcare/android/fragments/SelectInstallModeFragment.java

```diff
@@ -69,6 +94,65 @@ public void onClick(View v) {
             }
         });
 
+        SquareButtonWithText installFromLocal = (SquareButtonWithText)view.findViewById(R.id.btn_fetch_hub);
+        installFromLocal.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                Activity currentActivity = getActivity();
+                if (currentActivity instanceof CommCareSetupActivity) {
+                    showLocalAppDialog();
+                }
+            }
+        });
+
+        mFetchHubContainer = view.findViewById(R.id.btn_fetch_hub_container);
+
         return view;
     }
+
+    private void showLocalAppDialog() {
+        ContextThemeWrapper wrapper = new ContextThemeWrapper(getContext(), R.style.DialogBaseTheme);
+        final PaneledChoiceDialog chooseApp = new PaneledChoiceDialog(wrapper,
+                Localization.get("install.choose.local.app"));
+
+        DialogChoiceItem[] items = new DialogChoiceItem[mLocalApps.size()];
+        int count = 0;
+        for (final String[] app : mLocalApps) {
+            DialogChoiceItem item = new DialogChoiceItem(app[0], -1, new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    Activity currentActivity = getActivity();
+                    if (currentActivity instanceof CommCareSetupActivity) {
+                        ((CommCareSetupActivity)currentActivity).onURLChosen(app[1]);
+                    }
+                    chooseApp.dismiss();
+                }
+            });
+            items[count] = item;
+            count++;
+        }
+        chooseApp.setChoiceItems(items);
+        chooseApp.show();
+    }
+
+    @Override
+    public synchronized void onMicronodeDiscovery() {
+        boolean appsAvailable = false;
+        mLocalApps = new ArrayList<>();
+        for (MicroNode node : NsdTools.getAvailableMicronodes()) {
+            for (String[] applications : node.getAvailableApplications()) {
+                mLocalApps.add(applications);
+                appsAvailable = true;
+            }
+        }
+        Activity activity = getActivity();
+        if (appsAvailable && activity != null) {
+            getActivity().runOnUiThread(new Runnable() {
+                @Override
+                public void run() {
+                    mFetchHubContainer.setVisibility(View.VISIBLE);
```

Comment:
Moving this line into a helper method (like `showLocalInstallButton()`) would help a bit with code readability